### PR TITLE
ISPN-8090 Functional commands don't support Data conversions In Transactional Caches

### PR DESCRIPTION
**Pull request by rvansa**
* Remove some generics in functional commands that don't make sense;
  the function generics no longer match to the stored keys and values
* Add DataConversion to Mutations
* Use DataConversion.writeTo/readFrom directly

https://issues.jboss.org/browse/ISPN-8090

### CODE REVIEW

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java [hunk 1]
```diff
@@ -1328,6 +1328,15 @@ private void processResponse(org.jgroups.Address src, byte[] buffer, int offset,
          requests.addResponse(requestId, address, response);
       } catch (Throwable t) {
          log.errorProcessingResponse(requestId, src);
+         Exception e = t instanceof Exception ? ((Exception) t) : new CacheException(t);
+         try {
+            marshaller.objectFromByteBuffer(buffer, offset, length);
+         } catch (IOException e1) {
+            e1.printStackTrace();  // TODO: Customise this generated block
```
Comment (karesti) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java]: Don't forget the TODOs here !

File: core/src/main/java/org/infinispan/commands/functional/Mutations.java [hunk 2]
```diff
@@ -18,46 +20,72 @@ private Mutations() {}
...
       byte type = mutation.type();
       output.writeByte(type);
       switch (type) {
          case ReadWrite.TYPE:
             output.writeObject(((ReadWrite<K, V, ?>) mutation).f);
             break;
          case ReadWriteWithValue.TYPE:
-            output.writeObject(((ReadWriteWithValue<K, V, R>) mutation).value);
-            output.writeObject(((ReadWriteWithValue<K, V, R>) mutation).f);
+            ReadWriteWithValue<K, V, R> rwwv = (ReadWriteWithValue<K, V, R>) mutation;
```
Comment (karesti) [inline core/src/main/java/org/infinispan/commands/functional/Mutations.java]: Is there any way to refactor the code so we just need to do an if ? instead of checking every type to write f everytime etc ...

**Comment by karesti** [comment]
@rvansa checkstyle seems ko

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java (hunk 1 above)
Comment (rvansa) [reply core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java]: Ouch, thanks - this was a debugging code ofc :-/

File: core/src/main/java/org/infinispan/commands/functional/Mutations.java (hunk 2 above)
Comment (rvansa) [reply core/src/main/java/org/infinispan/commands/functional/Mutations.java]: Do you want me to remove the switch? I  could add `writeTo` method to each mutation, instead of calling `.type()`, but I prefer to keep the writeTo/readFrom symmetric (and readFrom cannot be refactored that way).

Comment (karesti) [reply core/src/main/java/org/infinispan/commands/functional/Mutations.java]: It's not a big deal for me, code seemed to me very similar in the switch so I suggested a change, but do whatever you think is better/homogenous for this case.

**Comment by rvansa** [comment]
Wait a sec, I've noticed that I've omitted few places..

**Comment by rvansa** [comment]
Ok, I've fixed several other generics (mostly the CommandFactory and xMapImpls)

**Comment by wburns** [comment]
Restarting CI build as it CI itself blew chunks on last run.

File: core/src/main/java/org/infinispan/commands/functional/Mutation.java [hunk 3]
```diff
@@ -1,11 +1,13 @@
...
 
+import org.infinispan.commands.functional.functions.InjectableComponent;
+import org.infinispan.encoding.DataConversion;
 import org.infinispan.functional.EntryView;
 
 /**
  * Simplified version of functional command used for read-only operations after transactional modifications.
  */
-public interface Mutation<K, V, R> {
+public interface Mutation<K, V, R> extends InjectableComponent {
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/functional/Mutation.java]: Hrmm I am not a fan of having this at this level. `InjectableComponent` to me is more of an implementation specific thing.

Also should help perf a tiny little bit, since some implementations don't need injection to be called.

File: core/src/main/java/org/infinispan/commands/functional/TxReadOnlyKeyCommand.java [hunk 4]
```diff
@@ -59,11 +59,22 @@ public void readFrom(ObjectInput input) throws IOException, ClassNotFoundExcepti
    }
 
    @Override
+   public void init(ComponentRegistry componentRegistry) {
+      super.init(componentRegistry);
+      // This may be called from parent's constructor when mutations are not initialized yet
```
Comment (wburns) [inline core/src/main/java/org/infinispan/commands/functional/TxReadOnlyKeyCommand.java]: Sorry what implications does this then have? It seems the `mutations` collection is not thread safe, so can we get `ConcurrentModificationException`?

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java [hunk 5]
```diff
@@ -1328,6 +1328,8 @@ private void processResponse(org.jgroups.Address src, byte[] buffer, int offset,
          requests.addResponse(requestId, address, response);
       } catch (Throwable t) {
          log.errorProcessingResponse(requestId, src);
+         Exception e = t instanceof Exception ? ((Exception) t) : new CacheException(t);
```
Comment (wburns) [inline core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java]: This change seems unrelated? What is this a fix for?

File: core/src/main/java/org/infinispan/commands/functional/Mutation.java (hunk 3 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/commands/functional/Mutation.java]: @wburns I agree, but OTOH the `org.infinispan.commands.functional` package isn't `@public` either.

File: core/src/main/java/org/infinispan/commands/functional/TxReadOnlyKeyCommand.java (hunk 4 above)
Comment (danberindei) [reply core/src/main/java/org/infinispan/commands/functional/TxReadOnlyKeyCommand.java]: @wburns I think it's just that on remote nodes the mutations list is initialized before `init()`, and on on the originator the mutations are added (already initialized) after `init()`. From a thread-safety perspective, AFAICT the collection is unique the command, and we never have 2 threads accessing the same command -- or we have a bug.

File: core/src/main/java/org/infinispan/commands/functional/Mutations.java [hunk 6]
```diff
@@ -89,15 +122,21 @@ public byte type() {
 
       @Override
       public R apply(EntryView.ReadWriteEntryView<K, V> view) {
-         return f.apply(value, view);
+         return f.apply((V) valueDataConversion.fromStorage(value), view);
+      }
+
+      @Override
+      public void inject(ComponentRegistry registry) {
+         registry.wireDependencies(valueDataConversion);
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/commands/functional/Mutations.java]: I only used `wireDependency` for `CacheTopologyControlCommand` because I knew those are pretty rare, I don't think it's efficient enough to use multiple times per operation.

File: core/src/main/java/org/infinispan/commands/functional/TxReadOnlyKeyCommand.java (hunk 4 above)
Comment (rvansa) [reply core/src/main/java/org/infinispan/commands/functional/TxReadOnlyKeyCommand.java]: Mr. Berindei is correct, it's just about initialization order, we need the `if`.

File: core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java (hunk 5 above)
Comment (rvansa) [reply core/src/main/java/org/infinispan/remoting/transport/jgroups/JGroupsTransport.java]: I was seeing a testsuite stall due to a bug (during development) causing NPE. The bug has been fixed, so this is not technically needed...

**Comment by rvansa** [comment]
Rebased and updated. @danberindei see `BaseMutation.inject`, is this what you were looking for?

**Comment by rvansa** [comment]
@wburns I had to keep the `Mutation extends InjectableComponent` because the `inject` is called during command intialization (see `TxReadOnlyKeyCommand.init`). Otherwise I would have to add `@Inject` to mutation implementations and call `wireDependencies` on each.

**Comment by pruivo** [comment]
@rvansa checkstyle fails!

**Comment by galderz** [comment]
Needs rebase

**Comment by rvansa** [comment]
Uuh, I completely forgot this is still open!

**Comment by galderz** [comment]
Last CI build did not complete, fired off new build

**Comment by tristantarrant** [comment]
Requeuing, but it seems the build hangs during the core testsuite.

**Comment by rvansa** [comment]
@tristantarrant I've reproduced stuck build in counters, will check out. Seems that stuck builds are not a random issue anymore but sth worth investigation.

**Comment by rvansa** [comment]
@tristantarrant OK, CI looks good now (besides failure in biased scattered cache which I should investigate).

**Comment by galderz** [comment]
@rvansa Shall we keep PR open until you fix that? Or do you want to create a follow up JIRA for that?

**Comment by rvansa** [comment]
@galderz No, that's completely unrelated, my comment was just a rant.

**Comment by rvansa** [comment]
@galderz You seemed eager to merge this, anything missing?

**Comment by galderz** [comment]
Actually no, I forgot about it before doing CR1.
